algorithms/src/std_algorithms/numeric/Kokkos_AdjacentDifference.hpp

```diff
@@ -0,0 +1,295 @@
+  BinaryOperator m_op;
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const index_type i) const {
+    const auto& my_value = m_first[i];
+    if (i == 0) {
+      m_dest_view(i) = my_value;
+    } else {
+      const auto& left_value = m_first[i - 1];
+      m_dest_view(i)         = m_op(my_value, left_value);
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  StdAdjacentDiffItToViewFunctor(IteratorType first, DestViewType dest_view,
+                                 BinaryOperator op)
+      : m_first(first), m_dest_view(dest_view), m_op(op) {}
+};
+
+template <class ViewTypeFrom, class OutputIteratorType>
+struct StdAdjDiffCopyFunctor {
+  ViewTypeFrom m_view_from;
+  OutputIteratorType m_first_dest;
+
+  KOKKOS_INLINE_FUNCTION
+  StdAdjDiffCopyFunctor(const ViewTypeFrom view_from,
+                        OutputIteratorType first_dest)
+      : m_view_from(view_from), m_first_dest(first_dest) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(int i) const { m_first_dest[i] = m_view_from(i); }
+};
+
+// ------------------------------------------
+// adjacent_difference_impl
+// ------------------------------------------
+template <class ExecutionSpace, class InputIteratorType,
+          class OutputIteratorType, class BinaryOp>
+OutputIteratorType adjacent_difference_impl(const std::string& label,
+                                            const ExecutionSpace& ex,
+                                            InputIteratorType first_from,
+                                            InputIteratorType last_from,
+                                            OutputIteratorType first_dest,
+                                            BinaryOp bin_op) {
+  // checks
+  static_assert_random_access_and_accessible(ex, first_from, first_dest);
+  static_assert_iterators_have_matching_difference_type(first_from, first_dest);
+  expect_valid_range(first_from, last_from);
+
+  if (first_from == last_from) {
+    return first_dest;
+  }
+
+  // potentially, we could handle better the case where
+  // source interval and output interval are not the same.
+  // in that case, we don't need the auxiliary view.
+
+  // aliases
+  using value_type    = typename OutputIteratorType::value_type;
+  using aux_view_type = ::Kokkos::View<value_type*, ExecutionSpace>;
+  using functor1_t    = StdAdjacentDiffItToViewFunctor<InputIteratorType,
+                                                    aux_view_type, BinaryOp>;
+  using functor2_t = StdAdjDiffCopyFunctor<aux_view_type, OutputIteratorType>;
+
+  // run
+  const auto num_elements = last_from - first_from;
+  aux_view_type aux_view("aux_view", num_elements);
+  ::Kokkos::parallel_for(label,
+                         RangePolicy<ExecutionSpace>(ex, 0, num_elements),
+                         functor1_t(first_from, aux_view, bin_op));
+  ::Kokkos::parallel_for("copy",
+                         RangePolicy<ExecutionSpace>(ex, 0, num_elements),
+                         functor2_t(aux_view, first_dest));
```

Comment:
Actually, you don't need the auxiliary view.  The standard say the behavior is undefined if the input and output ranges overlap in any way.
https://eel.is/c++draft/adjacent.difference#8

Reply:
you are referring to this remark right? "For the overloads with no ExecutionPolicy, result may be equal to first. For the overloads with an ExecutionPolicy, the ranges [first, last) and [result, result + (last - first)) shall not overlap." 
So my follow up question is: if a user passes `Kokkos:Serial`, which in my view is like similar to std serial, should we then allow the ranges to overlap?

Reply:
ok, changing this to have overlapping always UB